src/java/org/apache/cassandra/service/accord/AccordService.java

```diff
@@ -377,24 +371,33 @@ public synchronized void startup()
         node.commandStores().restoreShardStateUnsafe(topology -> configService.reportTopology(topology, true, true));
         configService.start();
 
-        long minEpoch = fetchMinEpoch();
-        if (minEpoch >= 0)
+        try
         {
-            for (long epoch = minEpoch; epoch <= metadata.epoch.getEpoch(); epoch++)
-                node.configService().fetchTopologyForEpoch(epoch);
+            // Fetch topologies up to current
+            List<Topology> topologies = fetchTopologies(0, metadata);
+            for (Topology topology : topologies)
+                configService.reportTopology(topology);
 
-            try
-            {
-                epochReady(metadata.epoch).get(DatabaseDescriptor.getTransactionTimeout(MILLISECONDS), MILLISECONDS);
-            }
-            catch (InterruptedException e)
-            {
-                throw new UncheckedInterruptedException(e);
-            }
-            catch (ExecutionException | TimeoutException e)
+            ClusterMetadataService.instance().log().addListener(configService.listener);
+            ClusterMetadata next = ClusterMetadata.current();
+
+            // if metadata was updated before we were able to add a listener, fetch remaining topologies
+            if (metadata != next)
             {
-                throw new RuntimeException(e);
+                topologies = fetchTopologies(metadata.epoch.getEpoch(), next);
+                for (Topology topology : topologies)
+                    configService.reportTopology(topology);
             }
+
+            epochReady(metadata.epoch).get();
+        }
+        catch (InterruptedException e)
+        {
+            throw new UncheckedInterruptedException(e);
```

Comment:
```suggestion
            Thread.currentThread().interrupt();
            throw new UncheckedInterruptedException(e);
```

since we don't handle, we just rethrow we should reset the interrupt flag so it doesn't get swallowed